### PR TITLE
WIP: support spark-3.4

### PR DESCRIPTION
**Pull request by zhouyuan**
## What changes were proposed in this pull request?

Support for Spark-3.4
`mvn clean package -Pbackends-velox -Prss -Pspark-3.4 -DskipTests`

(Fixes: \#ISSUE-ID)

## How was this patch tested?

UT for Spark-3.4 to be added

### CODE REVIEW

File: gluten-core/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala [hunk 1]
```diff
@@ -564,7 +558,11 @@ object FilterHandler {
               getLeftFilters(scan.dataFilters, flattenCondition(plan.condition))
             val newPartitionFilters =
               ExpressionConverter.transformDynamicPruningExpr(scan.partitionFilters, reuseSubquery)
-            new BatchScanExecTransformer(batchScan.output, scan, leftFilters ++ newPartitionFilters)
+            new BatchScanExecTransformer(
+              batchScan.output,
+              scan,
+              leftFilters ++ newPartitionFilters,
+              batchScan.table)
```
Comment (zhouyuan) [inline gluten-core/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala:565]: "batchScan.table" should move to shim layer

File: gluten-core/src/main/scala/io/glutenproject/execution/BatchScanExecTransformer.scala [hunk 2]
```diff
@@ -43,8 +45,22 @@ class BatchScanExecTransformer(
...
+  extends BatchScanExecShim(
+    output,
+    scan,
+    runtimeFilters,
+    keyGroupedPartitioning,
+    ordering,
+    table,
+    commonPartitionValues,
+    applyPartialClustering,
+    replicatePartitions)
```
Comment (zhouyuan) [inline gluten-core/src/main/scala/io/glutenproject/execution/BatchScanExecTransformer.scala:63]: should move `BatchScanExecShim` to shim layer

File: shims/common/src/main/scala/org/apache/spark/sql/execution/stat/StatFunctions.scala [hunk 3]
```diff
@@ -1,356 +0,0 @@
-/*
```
Comment (zhouyuan) [inline shims/common/src/main/scala/org/apache/spark/sql/execution/stat/StatFunctions.scala:1]: should move to shim layer

File: shims/spark34/src/main/scala/org/apache/spark/sql/catalyst/expressions/PromotePrecision.scala [hunk 4]
```diff
@@ -0,0 +1,37 @@
+/*
```
Comment (zhouyuan) [inline shims/spark34/src/main/scala/org/apache/spark/sql/catalyst/expressions/PromotePrecision.scala:1]: `PromotePrecision `was removed from Spark 3.4

**Comment by zhouyuan** [comment]
@JkSelf  this patch tends to add support for Spark 3.4 - the line of changes is not very big

CC: @zzcclp

File: shims/spark34/src/main/scala/org/apache/spark/sql/execution/datasources/v2/velox/DwrfScanBuilder.scala [hunk 5]
```diff
@@ -0,0 +1,44 @@
+/*
```
Comment (JkSelf) [inline shims/spark34/src/main/scala/org/apache/spark/sql/execution/datasources/v2/velox/DwrfScanBuilder.scala:1]: @zhouyuan We have removed DwrfScan. Why need in Spark3.4?

File: gluten-core/src/main/scala/org/apache/spark/rpc/GlutenDriverEndpoint.scala [hunk 6]
```diff
@@ -40,6 +40,8 @@ class GlutenDriverEndpoint extends IsolatedRpcEndpoint with Logging {
   private val driverEndpoint: RpcEndpointRef =
     rpcEnv.setupEndpoint(GlutenRpcConstants.GLUTEN_DRIVER_ENDPOINT_NAME, this)
 
+  // TODO(yuan): get thread cnt from spark context
+  override def threadCount(): Int = 1
```
Comment (JkSelf) [inline gluten-core/src/main/scala/org/apache/spark/rpc/GlutenDriverEndpoint.scala:44]: @zhouyuan Why need this change both in GlutenDriverEndpoint and GluenExecutroEndpoint?

File: shims/spark34/src/main/scala/org/apache/spark/sql/execution/datasources/v2/velox/DwrfScanBuilder.scala (hunk 5 above)
Comment (zhouyuan) [reply shims/spark34/src/main/scala/org/apache/spark/sql/execution/datasources/v2/velox/DwrfScanBuilder.scala:1]: yes, it should be removed. it's accidently left during my rebase work.

File: gluten-core/src/main/scala/org/apache/spark/rpc/GlutenDriverEndpoint.scala (hunk 6 above)
Comment (zhouyuan) [reply gluten-core/src/main/scala/org/apache/spark/rpc/GlutenDriverEndpoint.scala:44]: `IsolatedRpcEndpoint ` introduced a virtual threadCount func

**Comment by zhouyuan** [comment]
replaced by #3360

**Pull request closed without merging**